src/Bicep.Decompiler/TemplateConverter.cs

```diff
@@ -0,0 +1,907 @@
+                    SyntaxHelpers.CreateToken(TokenType.RightParen, ")"));
+            }
+
+            foreach (var property in expression.Properties)
+            {
+                if (property is JTokenExpression jTokenExpression && jTokenExpression.Value.Type == JTokenType.String)
+                {
+                    // TODO handle special chars and generate array access instead
+                    baseSyntax = new PropertyAccessSyntax(
+                        baseSyntax,
+                        SyntaxHelpers.CreateToken(TokenType.Dot, "."),
+                        SyntaxHelpers.CreateIdentifier(jTokenExpression.Value.Value<string>()!));
+                }
+                else
+                {
+                    baseSyntax = new ArrayAccessSyntax(
+                        baseSyntax,
+                        SyntaxHelpers.CreateToken(TokenType.LeftSquare, "["),
+                        ParseLanguageExpression(property),
+                        SyntaxHelpers.CreateToken(TokenType.RightSquare, "]"));
+                }
+            }
+
+            return baseSyntax;
+        }
+
+        private SyntaxBase ParseLanguageExpression(LanguageExpression expression)
+            => expression switch
+            {
+                JTokenExpression jTokenExpression => ParseJTokenExpression(jTokenExpression),
+                FunctionExpression functionExpression => ParseFunctionExpression(functionExpression),
+                _ => throw new NotImplementedException($"Unrecognized expression {ExpressionsEngine.SerializeExpression(expression)}"),
+            };
+
+        private SyntaxBase ParseString(string? value)
+        {
+            if (value == null)
+            {
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            if (ExpressionsEngine.IsLanguageExpression(value))
+            {
+                var expression = ExpressionsEngine.ParseLanguageExpression(value);
+
+                return ParseLanguageExpression(expression);
+            }
+            
+            return SyntaxHelpers.CreateStringLiteral(value);
+        }
+
+        private static SyntaxBase ParseIntegerJToken(JValue value)
+        {
+            // JTokenType.Integer can encapsulate non-C#-int numbers.
+            // Best we can do in this case is to convert them to strings.
+            var longValue = value.Value<long>();
+            if (longValue < int.MinValue || longValue > int.MaxValue)
+            {
+                return SyntaxHelpers.CreateStringLiteral(value.ToString());
+            }
+
+            return new NumericLiteralSyntax(SyntaxHelpers.CreateToken(TokenType.Number, value.ToString()), (int)longValue);
+        }
+
+        private SyntaxBase ParseJValue(JValue value)
+            => value.Type switch {
+                JTokenType.String => ParseString(value.ToString()),
```

Comment:
Some of the template function implementations used the internal `IsTextBasedJTokenType` helper, which also checks for `JTokenType.Uri`. Do we need that? Are there also other `JTokenType` values that we need to handle?

Reply:
As far as I could tell from reading the Newtonsoft code, Uri or any of the other specialized token types won't be returned on via `JToken.Parse`. I also looked through the output of the quickstart repo conversion (600+ templates), and couldn't find any unexpected instances of the below exception being thrown.

Reply:
Makes sense.